File: go/gain.go
package main

import "math"

type Counter map[float64]float64

func frequency(dataset []Sample, attr string) Counter {
	counter := make(Counter)
	for _, s := range dataset {
		value := s.GetValue(attr)

		v, ok := counter[value]
		if !ok {
			counter[value] = 1.0

			continue
		}

		counter[value] = v + 1.0
	}
	return counter
}

func logN(x float64, n float64) float64 {
	return math.Log(x) / math.Log(n)
}

func entropy(dataset []Sample, attr string) float64 {
	counter := frequency(dataset, attr)

	var (
		n int     = 2
		s float64 = 0
		e float64 = 0
	)

	if len(dataset) > n {
		n = len(dataset)
	}

	for _, value := range counter {
		s += value
	}

	for _, value := range counter {
		e += (value / s) * logN(value/s, float64(n))
	}

	return e * -1
}

func getSubset(dataset []Sample, attr string, value float64) []Sample {
	s := make([]Sample, 0)
	for _, sample := range dataset {
		if sample.GetValue(attr) == value {
			s = append(s, sample)
		}
	}
	return s
}

func gain(dataset []Sample, attr, classAttr string) float64 {
	var (
		s float64 = 0.0
		e float64 = 0.0
	)

	values := frequency(dataset, attr)

	for _, value := range values {
		s += value
	}

	for key, value := range values {
		prob := value / s
		subset := getSubset(dataset, attr, key)
		e += prob * entropy(subset, classAttr)
	}
	return entropy(dataset, classAttr) - e
}

File: python/classifier/gain.py
from collections import Counter
from math import log


def _frequency(dataset, attr):
    counter = Counter()
    for sample in dataset:
        k = getattr(sample, attr)
        counter[k] += 1.0
    return counter


def _entropy(data, attr):
    frequency = _frequency(data, attr)
    s = float(sum(f for f in frequency.values()))
    n = max(2, len(frequency))
    return -sum((f / s) * log(f / s, n) for f in frequency.values() if f)


def gain(dataset, attr, class_attr):
    subset_entropy = 0.0
    frequency = _frequency(dataset, attr)

    for value in frequency.keys():
        subset = []
        for sample in dataset:
            if getattr(sample, attr) == value:
                subset.append(sample)

        e = _entropy(subset, class_attr)
        prob = frequency[value] / sum(frequency.values())
        subset_entropy += prob * e

    main_entropy = _entropy(dataset, class_attr)
    return main_entropy - subset_entropy
